source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,192 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+
+class ClusterCallbacks : public ClusterUpdateCallbacks {
+public:
+  ClusterCallbacks(ClusterAddedCb cb) : cb_(std::move(cb)) {}
+
+  void onClusterAddOrUpdate(ThreadLocalCluster& cluster) override { cb_(cluster); };
+
+  void onClusterRemoval(const std::string&) override {}
+
+private:
+  ClusterAddedCb cb_;
+};
+
+} // namespace
+
+ClusterManagerImpl::ClusterDiscoveryManager::ClusterDiscoveryManager(
+    ThreadLocalClusterManagerImpl& parent)
+    : parent_(parent) {}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::ensureLifecycleCallbacksAreInstalled() {
+  if (callbacks_handle_) {
+    return;
+  }
+  callbacks_ = std::make_unique<ClusterCallbacks>([this](ThreadLocalCluster& cluster) {
+    processClusterName(cluster.info()->name(), ClusterDiscoveryStatus::Available);
+  });
+  callbacks_handle_ = parent_.parent_.addThreadLocalClusterUpdateCallbacks(*callbacks_);
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::processClusterName(
+    const std::string& name, ClusterDiscoveryStatus cluster_status) {
+  // Extracting the list of callbacks from the map makes resetting the
+  // handle inside the callback safe, because handle would try to find
+  // the list of callbacks in the map and would find nothing, instead
+  // of removing an item from the list while iterating the list.
+  auto map_node_handle = pending_clusters_.extract(name);
+  if (map_node_handle.empty()) {
+    return;
+  }
+  for (const auto& weak_callback : map_node_handle.mapped()) {
+    auto callback = weak_callback.lock();
+    if (callback != nullptr) {
+      (*callback)(cluster_status);
+    }
+  }
+  maybePostResetCallbacks();
+}
+
+ClusterManagerImpl::ClusterDiscoveryManager::Pair
+ClusterManagerImpl::ClusterDiscoveryManager::addCallback(
+    const std::string& name, const ClusterDiscoveryCallbackSharedPtr& callback) {
+  ensureLifecycleCallbacksAreInstalled();
+
+  auto& callbacks_list = pending_clusters_[name];
+  auto it = callbacks_list.emplace(callbacks_list.end(), callback);
+  auto handle = std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, name, it);
+  auto discovery_in_progress = (callbacks_list.size() > 1);
+  return {std::move(handle), discovery_in_progress};
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::erase(const std::string& name,
+                                                        CallbackListIterator it) {
+  const bool drop_list = eraseFromList(name, it);
+  if (drop_list) {
+    pending_clusters_.erase(name);
+  }
+  maybePostResetCallbacks();
+}
+
+bool ClusterManagerImpl::ClusterDiscoveryManager::eraseFromList(const std::string& name,
+                                                                CallbackListIterator it) {
+  auto map_it = pending_clusters_.find(name);
+  if (map_it == pending_clusters_.end()) {
+    return false;
+  }
+  auto& list = map_it->second;
+  list.erase(it);
+  return list.empty();
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::maybePostResetCallbacks() {
```

Comment:
Sorry I'm a little confused on the purpose of this function. Can you clarify? I think you can just leave the callbacks handle installed the entire time? This is related to my other comment on the ensure function.